rdfind.1
.\" View this file with
.\" groff -man -Tutf8 rdfind.1 |less
.\"
.\" Author Paul Dreik 2006
.\" see LICENSE for details.
.TH rdfind "1" 1.8.0 "Feb 2026" rdfind
.SH NAME
rdfind \- finds duplicate files
.SH SYNOPSIS
.B rdfind [ options ]
.I directory1 | file1
.B [
.I directory2 | file2
.B ] ...
.SH DESCRIPTION
.B rdfind
finds duplicate files across and/or within several directories.  It
calculates checksums only if necessary. Directories are searched
recusively.

If two (or more) equal files are found, the program decides which of
them is the original and the rest are considered duplicates. This
is done by ranking the files to each other and deciding which has the
highest rank. See section RANKING for details.

By default, no action is taken besides creating a file with the
detected files and showing the possible amount of saved space.

If you need better control over the ranking than given, you can use
some preprocessor which sorts the file names in desired order and then
run the program using xargs. See examples below for how to use find
and xargs in conjunction with rdfind.

To include files or directories that have names starting with \-, use
rdfind ./\- to not confuse them with options.

.SH RANKING
Given two or more equal files, the one with the highest rank is
selected to be the original and the rest are duplicates. The rules of
ranking are given below, where the rules are executed from start until
an original has been found. Given two files A and B which have equal
size and content, the ranking is as follows:

If A was found while scanning an input argument earlier than B, A
is higher ranked.

If A was found at a directory depth lower than B, A is higher ranked
(A is closer to the root).

if A and B are found during scanning of the same input argument and share
the same directory depth, the one that ranks highest depends on if
deterministic operation is enabled. This is on by default, see option
\fB\-deterministic\fR). If enabled, which one ranks highest is
unspecified but deterministic. If disabled, the one that was reported
first from the file system is highest ranked.

.SH OPTIONS
Searching options etc:
.TP
.BR \-ignoreempty " " \fItrue\fR|\fIfalse\fR
Ignore empty files. Setting this to true (the default) is equivalent to
\-minsize 1, false is equivalent to \-minsize 0.
.TP
.BR \-minsize " "\fIN\fR
Ignores files with less than N bytes. Default is 1, meaning empty files
are ignored.
.TP
.BR \-maxsize " "\fIN\fR
Ignores files with N bytes or more. Default is 0, which means this check
is disabled.
.TP
.BR \-followsymlinks " " \fItrue\fR|\fIfalse\fR
Follow symlinks. Default is false.
.TP
.BR \-removeidentinode " " \fItrue\fR|\fIfalse\fR
Removes items found which have identical inode and device ID. Default
is true.
.TP
.BR \-checksum " " \fInone\fR|\fImd5\fR|\fIsha1\fR|\fIsha256\fR|\fIsha512|\fIxxh128\fR
What type of checksum to be used: md5, sha1, sha256, sha512 or xxh128. The default is
sha1 since version 1.4.0. xxh128 is a very fast checksum, but not of cryptographic
quality. xxh support is optional and requires that rdfind was configured with
--with-xxhash. In case xxh is used but there is no support, an error is returned.
Checksum none can be used to skip checksumming altogether. \fBThis is not recommended!\fR
In case files of the same size have contents that differ it is likely they are falsely
consider duplicates, leading to file removal (depending on other options).
.TP
.BR \-buffersize " " \fIN\fR
Chunksize in bytes when calculating the checksum
for files, smaller or bigger can improve performance
dependent on filesystem and checksum algorithm.
The default is 1 MiB, the maximum allowed is 128MiB (inclusive).
.TP
.BR \-firstbytessize " " \fIN\fR
Size in bytes when scanning the first bytes of each file, prior to full
checksumming. Setting this to 0 means skipping the step entirely.
.TP
.BR \-lastbytessize " " \fIN\fR
Size in bytes when scanning the last bytes of each file, prior to full
checksumming. Setting this to 0 means skipping the step entirely.
.TP
.BR \-deterministic " " \fItrue\fR|\fIfalse\fR
If set (the default), sort files of equal rank in an unspecified but
deterministic order. This makes the behaviour independent of in which
order files are listed when querying the file system.
.PP
Action options:
.TP
.BR \-makesymlinks " " \fItrue\fR|\fIfalse\fR
Replace duplicate files with symbolic links. Default is false.
.TP
.BR \-makehardlinks " " \fItrue\fR|\fIfalse\fR
Replace duplicate files with hard links. Default is false.
.TP
.BR \-makeresultsfile " " \fItrue\fR|\fIfalse\fR
Make a results file in the current directory. Default is true. If the
file exists, it is overwritten. This does not affect whether items are
deleted. See \-dryrun for how to disable deletions.
.TP
.BR \-outputname " " \fIname\fR
Make the results file name to be "name" instead of the default
results.txt.
.TP
.BR \-deleteduplicates " " \fItrue\fR|\fIfalse\fR
Delete (unlink) files. Default is false.
.PP
General options:
.TP
.BR \-progress " " \fItrue\fR|\fIfalse\fR
Show progress during elimination. Defaults to false.
.TP
.BR \-sleep " " \fIX\fRms
Sleeps X milliseconds between reading each file, to reduce
load. Default is 0 (no sleep). Note that only a few values are
supported at present: 0,1-5,10,25,50,100 milliseconds.
.TP
.BR \-n ", " \-dryrun " " \fItrue\fR|\fI(false)\fR
By default, rdfind does nothing except creating a results file.  In
case one of the actions flags like -deleteduplicates is set, dryrun
can be set to let rdfind print what it would do, instead of actually
doing it. The default is false.
.TP
.BR \-h ", " \-help ", " \-\-help
Displays a brief help message.
.TP
.BR \-v ", " \-version ", " \-\-version
Displays the version number.
.SH EXAMPLES
.TP
Search for duplicate files in the home directory and a backup directory:
.B rdfind ~ /mnt/backup
.TP
Delete duplicates in a backup directory:
.B rdfind \-deleteduplicates true /mnt/backup
.TP
Search for duplicate files in directories called foo:
.B find . \-type d \-name foo \-print0 |xargs \-0 rdfind
.SH FILES
.I results.txt
(the default name is results.txt and can be changed with option outputname,
see above) The results file results.txt will contain one row per duplicate file
found, along with a header row explaining the columns.
A text describes why the file is considered a duplicate:

DUPTYPE_UNKNOWN some internal error

DUPTYPE_FIRST_OCCURRENCE the file that is considered to be the original.

DUPTYPE_WITHIN_SAME_TREE files in the same tree (found when processing
the directory in the same input argument as the original)

DUPTYPE_OUTSIDE_TREE the file is found during processing another input
argument than the original.
.SH ENVIRONMENT
.SH DIAGNOSTICS
.SH EXIT VALUES
0 on success, nonzero otherwise.
.SH BUGS/FEATURES
When specifying the same directory twice, it keeps the first
encountered as the most important (original), and the rest as
duplicates. This might not be what you want.

The symlink creates absolute links. This might not be what you
want. To create relative links instead, you may use the symlinks(8)
command, which is able to convert absolute links to relative links.

Older versions unfortunately contained a misspelling on the word
occurrence. This is now corrected (since 1.3), which might affect
user scripts parsing the output file written by rdfind.

.SH SECURITY CONSIDERATIONS
Avoid manipulating the directories while rdfind is reading.
rdfind is quite brittle in that case. Especially, when deleting
or making links, rdfind can be subject to a symlink attack.
Use with care!

Rfind was written with the intention of deduplicating directories
owned by the same user. It does not handle file ownership and
permissions. It is not recommended to run it across directories with
several owners.

.SH AUTHOR
Paul Dreik 2006-2026, reachable at rdfind@pauldreik.se
Rdfind can be found at https://rdfind.pauldreik.se/

Do you find rdfind useful? Drop me a line! It is always fun to
hear from people who actually use it and what data collections
they run it on.
.SH THANKS
Several persons have helped with suggestions and improvements:
Niels Möller, Carl Payne and Salvatore Ansani. Thanks also to you
who tested the program and sent me feedback.
.SH VERSION
1.8.0 (release date 2026-02-22)
.SH COPYRIGHT
This program is distributed under GPLv2 or later, at your option.
.SH "SEE ALSO"
.BR md5sum,
.BR sha1sum,
.BR find,
.BR symlinks


NEWS
next
optionally disable the checksum step by giving -checksum none
optionally show progress
optionally adjust the size of first/last bytes, or disable it completely.
1.7.0
requires a C++17 capable compiler.
new fast non-cryptographic hash xxh
controllable buffer size for calculating checksum
1.6.0
supports sha512
1.4.0alpha
now requires a c++11 capable compiler
1.3.5
Capitalization of bytes is now more correct in the console output.
It is unlikely this affects usage, unless the user actually parses the
output of the command rather than the results file.

sha1 was always selected over md5, even if the user specified md5
on the commandline. This is now fixed.
1.3.1
now builds with gcc 4.7
1.3.0
Spelling of OCCURRENCE is now correct. This might affect user scripts
parsing the output file (results.txt). 
1.2.3+
fixed bug with failure to read files given directly on command line
added debug support in configure script




rdfind.cc
/**
   copyright 2006-2017 Paul Dreik (earlier Paul Sundvall)
   Distributed under GPL v 2.0 or later, at your option.
   See LICENSE for further details.
*/

#include "config.h" //header file from autoconf

static_assert(__cplusplus >= 201703L,
              "this code requires a C++17 capable compiler!");

// std
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

// project
#include "CmdlineParser.hh"
#include "Dirlist.hh"     //to find files
#include "Fileinfo.hh"    //file container
#include "Options.hh"     //
#include "RdfindDebug.hh" //debug macro
#include "Rdutil.hh"      //to do some work

// global variables

// this vector holds the information about all files found
std::vector<Fileinfo> filelist;
const Options* global_options{};

/**
 * this contains the command line index for the path currently
 * being investigated. it has to be global, because function pointers
 * are being used.
 */
int current_cmdline_index = 0;

// function to add items to the list of all files
static int
report(const std::string& path, const std::string& name, int depth)
{

  RDDEBUG("report(" << path.c_str() << "," << name.c_str() << "," << depth
                    << ")" << std::endl);

  // expand the name if the path is nonempty
  std::string expandedname = path.empty() ? name : (path + "/" + name);

  Fileinfo tmp(std::move(expandedname), current_cmdline_index, depth);
  if (tmp.readfileinfo()) {
    if (tmp.isRegularFile()) {
      const auto size = tmp.size();
      if (size >= global_options->minimumfilesize &&
          size < global_options->maximumfilesize) {
        filelist.emplace_back(std::move(tmp));
      }
    }
  } else {
    std::cerr << "failed to read file info on file \"" << tmp.name() << "\"\n";
    return -1;
  }
  return 0;
}

int
main(int narg, const char* argv[])
{
  if (narg == 1) {
    usage();
    return 0;
  }

  // parse the input arguments
  Parser parser(narg, argv);

  const Options o = parseOptions(parser);

  // set the dryrun string
  const std::string dryruntext(o.dryrun ? "(DRYRUN MODE) " : "");

  // an object to do sorting and duplicate finding
  Rdutil gswd(filelist);

  // an object to traverse the directory structure
  Dirlist dirlist(o.followsymlinks);

  // this is what function is called when an object is found on
  // the directory traversed by walk. Make sure the pointer to the
  // options is set as well.
  global_options = &o;
  dirlist.setcallbackfcn(&report);

  // now loop over path list and add the files

  // done with arguments. start parsing files and directories!
  for (; parser.has_args_left(); parser.advance()) {
    // get the next arg.
    const std::string file_or_dir = [&]() {
      std::string arg(parser.get_current_arg());
      // remove trailing /
      while (arg.back() == '/' && arg.size() > 1) {
        arg.erase(arg.size() - 1);
      }
      return arg;
    }();

    auto lastsize = filelist.size();
    std::cout << dryruntext << "Now scanning \"" << file_or_dir << "\"";
    std::cout.flush();
    current_cmdline_index = parser.get_current_index();
    dirlist.walk(file_or_dir, 0);
    std::cout << ", found " << filelist.size() - lastsize << " files."
              << std::endl;

    // if we want deterministic output, we will sort the newly added
    // items on depth, then filename.
    if (o.deterministic) {
      gswd.sort_on_depth_and_name(lastsize);
    }
  }

  std::cout << dryruntext << "Now have " << filelist.size()
            << " files in total." << std::endl;

  // mark files with a number for correct ranking. The only ordering at this
  // point is that files found on early command line index are earlier in the
  // list.
  gswd.markitems();

  if (o.remove_identical_inode) {
    // remove files with identical devices and inodes from the list
    std::cout << dryruntext << "Removed " << gswd.removeIdenticalInodes()
              << " files due to nonunique device and inode." << std::endl;
  }

  std::cout << dryruntext << "Total size is " << gswd.totalsizeinbytes()
            << " bytes or ";
  gswd.totalsize(std::cout) << std::endl;

  std::cout << dryruntext << "Removed " << gswd.removeUniqueSizes()
            << " files due to unique sizes from list. ";
  std::cout << filelist.size() << " files left." << std::endl;

  // ok. we now need to do something stronger to disambiguate the duplicate
  // candidates. start looking at the contents.
  std::vector<std::pair<Fileinfo::readtobuffermode, const char*>> modes{
    { Fileinfo::readtobuffermode::NOT_DEFINED, "" },
  };
  if (o.first_bytes_size > 0) {
    modes.emplace_back(Fileinfo::readtobuffermode::READ_FIRST_BYTES,
                       "first bytes");
  }
  if (o.last_bytes_size > 0) {
    modes.emplace_back(Fileinfo::readtobuffermode::READ_LAST_BYTES,
                       "last bytes");
  }
  if (o.usemd5) {
    modes.emplace_back(Fileinfo::readtobuffermode::CREATE_MD5_CHECKSUM,
                       "md5 checksum");
  }
  if (o.usesha1) {
    modes.emplace_back(Fileinfo::readtobuffermode::CREATE_SHA1_CHECKSUM,
                       "sha1 checksum");
  }
  if (o.usesha256) {
    modes.emplace_back(Fileinfo::readtobuffermode::CREATE_SHA256_CHECKSUM,
                       "sha256 checksum");
  }
  if (o.usesha512) {
    modes.emplace_back(Fileinfo::readtobuffermode::CREATE_SHA512_CHECKSUM,
                       "sha512 checksum");
  }
  if (o.usexxh128) {
    modes.emplace_back(Fileinfo::readtobuffermode::CREATE_XXH128_CHECKSUM,
                       "xxh128 checksum");
  }

  std::function<void(std::size_t)> progress_callback;

  for (auto it = modes.begin() + 1; it != modes.end(); ++it) {
    std::cout << dryruntext << "Now eliminating candidates based on "
              << it->second << ": " << std::flush;

    if (o.showprogress) {
      progress_callback = []() {
        // format the total count only once, not each iteration.
        std::ostringstream oss;
        oss << "/" << filelist.size() << ")"
            << "\033[u"; // Restore the cursor to the saved position;
        return [suffix = oss.str()](std::size_t completed) {
          std::cout
            << "\033[s\033[K" // Save the cursor position & clear following text
            << "(" << completed << suffix << std::flush;
        };
      }();
    }

    // read bytes (destroys the sorting, for disk reading efficiency)
    gswd.fillwithbytes(it[0].first, it[-1].first, o, progress_callback);

    // remove non-duplicates
    std::cout << "removed " << gswd.removeUniqSizeAndBuffer()
              << " files from list. ";
    std::cout << filelist.size() << " files left." << std::endl;
  }

  // What is left now is a list of duplicates, ordered on size.
  // We also know the list is ordered on size, then bytes, and all unique
  // files are gone so it contains sequences of duplicates. Go ahead and mark
  // them.
  gswd.markduplicates();

  std::cout << dryruntext << "It seems like you have " << filelist.size()
            << " files that are not unique\n";

  std::cout << dryruntext << "Totally, ";
  gswd.saveablespace(std::cout) << " can be reduced." << std::endl;

  // traverse the list and make a nice file with the results
  if (o.makeresultsfile) {
    std::cout << dryruntext << "Now making results file " << o.resultsfile
              << std::endl;
    gswd.printtofile(o.resultsfile);
  }

  // traverse the list and replace with symlinks
  if (o.makesymlinks) {
    std::cout << dryruntext << "Now making symbolic links. creating "
              << std::endl;
    const auto tmp = gswd.makesymlinks(o.dryrun);
    std::cout << "Making " << tmp << " links." << std::endl;
    return 0;
  }

  // traverse the list and replace with hard links
  if (o.makehardlinks) {
    std::cout << dryruntext << "Now making hard links." << std::endl;
    const auto tmp = gswd.makehardlinks(o.dryrun);
    std::cout << dryruntext << "Making " << tmp << " links." << std::endl;
    return 0;
  }

  // traverse the list and delete files
  if (o.deleteduplicates) {
    std::cout << dryruntext << "Now deleting duplicates:" << std::endl;
    const auto tmp = gswd.deleteduplicates(o.dryrun);
    std::cout << dryruntext << "Deleted " << tmp << " files." << std::endl;
    return 0;
  }
  return 0;
}
